Added network-specific overrides to StreamRegistry

jtakalai opened the pull request:
set the gas price on Polygon to (currently rather high) 50gwei to see if that helps an urgent gas pricing issue

harbu commented on test/unit/tracker.test.js:
```diff
@@ -1,10 +1,55 @@
+        expect(tracker.getAddress()).toEqual('ws://127.0.0.1:30300')
+    })
+
+    it('tracker must check all required information for new incoming connection and not crash', async () => {
```
This is testing functionality of `WsEndpoint` right? Should these tests be in 'ws-endpoint.test.js' or some new test file dedicated to WsEndpoint?

mirpo replied:
Soon only tracker will use websocket

jtakalai committed:
feat: added network-specific overrides to StreamRegistry

set the gas price on Polygon to (currently rather high) 50gwei to see if that helps an urgent gas pricing issue

harbu commented on packages/client/src/StreamRegistry.ts:
```diff
@@ -181,7 +186,7 @@ export class StreamRegistry implements Context {
         let tx
         if (StreamIDUtils.isENSAddress(domain)) {
-            tx = await this.streamRegistryContract!.createStreamWithENS(domain, path, JSON.stringify(normalizedProperties))
+            tx = await this.streamRegistryContract!.createStreamWithENS(domain, path, JSON.stringify(normalizedProperties), ethersOverrides)
```
Is there a way to pass the `ethersOverrides` to the `streamRegistryContract` at once (e.g. constructor or setter) without having to repeat?

jtakalai replied:
I wrote it like that so that it could be overridden inside the function call too by adding it to the function signature if needed later...

harbu commented on packages/client/src/StreamRegistry.ts:
```diff
@@ -94,6 +95,8 @@ export class StreamRegistry implements Context {
         this.streamRegistryContractReadonly = new Contract(this.config.streamRegistryChainAddress,
             StreamRegistryArtifact, this.chainProvider) as StreamRegistryContract
+        const chainId = this.config.streamRegistryChainRPC?.chainId
+        this.defaultOverrides = this.config.ethereumNetworks?.find((network) => network.chainId === chainId)?.overrides
```
Do we need a check here in case `chainId` were undefined for whatever reason.
```suggestion
        if (chainId !== undefined) {
            this.defaultOverrides = this.config.ethereumNetworks?.find((network) => network.chainId === chainId)?.overrides
        }
```

jtakalai replied:
I guess the next line would still give out an undefined?

harbu commented on packages/client/src/ConfigBase.ts:
```diff
@@ -165,6 +165,16 @@ export const STREAM_CLIENT_DEFAULTS: StrictStreamrClientConfig = {
+
+    ethereumNetworks: [
+        {
+            chainId: 137,
```
Is there a reason to do it indirectly like this, instead of just having `overrides` property reside under config property `streamRegistryChainRPC`?

jtakalai replied:
I could put it there as well. Was thinking why not start writing stuff into a bit more "multi chain" structure, a bit like ETH-184 or https://hardhat.org/config/#networks-configuration and that array was just because it seemed simpler in terms of TypeScript types.

jtakalai replied:
I've changed it into network names as keys instead of array in #371

jtakalai committed:
fix: Overrides object

undefined doesn't work so well for the ethers library functions
